momaland/envs/multiwalker_stability/momultiwalker_stability_base.py

```diff
@@ -0,0 +1,228 @@
+
+        for i in range(self.n_walkers):
+            if self.walkers[i].hull is None:
+                obs.append(np.zeros_like(self.observation_space[i].low))
+                continue
+            pos = self.walkers[i].hull.position
+            x, y = pos.x, pos.y
+            xpos[i] = x
+
+            walker_obs = self.walkers[i].get_observation()
+            neighbor_obs = []
+            for j in [i - 1, i + 1]:
+                # if no neighbor (for edge walkers)
+                if j < 0 or j == self.n_walkers or self.walkers[j].hull is None:
+                    neighbor_obs.append(0.0)
+                    neighbor_obs.append(0.0)
+                else:
+                    xm = (self.walkers[j].hull.position.x - x) / self.package_length
+                    ym = (self.walkers[j].hull.position.y - y) / self.package_length
+                    neighbor_obs.append(self.np_random.normal(xm, self.position_noise))
+                    neighbor_obs.append(self.np_random.normal(ym, self.position_noise))
+            xd = (self.package.position.x - x) / self.package_length
+            yd = (self.package.position.y - y) / self.package_length
+            neighbor_obs.append(self.np_random.normal(xd, self.position_noise))
+            neighbor_obs.append(self.np_random.normal(yd, self.position_noise))
+            neighbor_obs.append(self.np_random.normal(self.package.angle, self.angle_noise))
+            obs.append(np.array(walker_obs + neighbor_obs))
+
+            shaping = -5.0 * abs(walker_obs[0])
+            rewards[i, 0] = shaping - self.prev_shaping[i]
+            self.prev_shaping[i] = shaping
+
+        package_shaping = self.forward_reward * 130 * self.package.position.x / SCALE
+        rewards[:, 0] = package_shaping - self.prev_package_shaping  # obj1: move forward
+        self.prev_package_shaping = package_shaping
+
+        self.scroll = xpos.mean() - VIEWPORT_W / SCALE / 5 - (self.n_walkers - 1) * WALKER_SEPERATION * TERRAIN_STEP
+
+        done = [False] * self.n_walkers
+        for i, (fallen, walker) in enumerate(zip(self.fallen_walkers, self.walkers)):
+            if fallen:
+                rewards[i, 0] += self.fall_reward
+                if self.remove_on_fall:
+                    walker._destroy()
+                if not self.terminate_on_fall:
+                    rewards[i, 0] += self.terminate_reward
+                done[i] = True
+        if (self.terminate_on_fall and np.sum(self.fallen_walkers) > 0) or self.game_over or self.package.position.x < 0:
+            rewards[:, 0] += self.terminate_reward
+            done = [True] * self.n_walkers
+        elif self.package.position.x > (self.terrain_length - TERRAIN_GRASS) * TERRAIN_STEP:
+            done = [True] * self.n_walkers
+
+        # package stability obj
+        pkg_angle_delta = abs(self.previous_pkg_angle - self.package.angle)
+        rewards[:, 1] = pkg_angle_delta * self.stability_reward
```

Comment:
I'd prefer restricting from using scaling factors in MORL since we normally don't need them.

Reply:
Then without the scale the bounds will be `+/- 0.0175`. That wouldn't be too small for using it?